src/util/strencodings.h

```diff
@@ -67,6 +68,67 @@ std::vector<Byte> ParseHex(std::string_view hex_str)
 {
     return TryParseHex<Byte>(hex_str).value_or(std::vector<Byte>{});
 }
+
+/** consteval version of HexDigit() without the lookup table. */
+consteval uint8_t ConstevalHexDigit(const char c)
+{
+    if (c >= '0' && c <= '9') return c - '0';
+    if (c >= 'a' && c <= 'f') return c - 'a' + 0xa;
+
+    throw "Only lowercase hex digits are allowed, for consistency";
+}
+
+struct ConstevalHexLiteral {
+    const std::string_view inner;
+    template <size_t N>
+    consteval ConstevalHexLiteral(const char (&hex_str)[N]) : ConstevalHexLiteral{std::string_view{hex_str}} {}
+    consteval ConstevalHexLiteral(const std::string_view hex_str) : inner{hex_str} { Validate(inner); }
+    consteval ConstevalHexLiteral(std::nullptr_t) = delete;
+
+    static consteval void Validate(const std::string_view hex_str)
+    {
+        if (hex_str.size() % 2 == 1) throw "2 hex digits required per byte";
+
+        std::ranges::for_each(hex_str, ConstevalHexDigit);
+    }
+};
+
+/** Validates input at compile time, but allocates the vector at runtime in
+ *  order to cross the compile time/runtime dynamic memory barrier.
+ */
+template <typename Byte = uint8_t>
+std::vector<Byte> VecFromHex(ConstevalHexLiteral hex_str)
+{
+    // No need to check HexDigit() return values or that hex_str has even length
+    // since ConstevalHexLiteral has already validated.
+    std::vector<Byte> rv(hex_str.inner.size() / 2);
+    auto it = hex_str.inner.begin();
+    for (auto& elem : rv) {
+        auto hi = HexDigit(*(it++)) << 4;
+        elem = Byte(hi | HexDigit(*(it++)));
+    }
+    return rv;
+}
+
+/** Converts from hex string literal to byte array at compile time.
+ * @warning Unlike VecFromHex which returns a vector, the returned array may use
+ *          significant stack space if called inside a function.
+ */
+template <typename Byte = uint8_t, size_t Size>
+consteval std::array<Byte, Size / 2> ArrayFromHex(const char (&hex_str)[Size])
+    // 2 hex digits required per byte + implicit null terminator
+    requires (Size % 2 == 1)
+{
+    if (hex_str[Size - 1] != '\0') throw "null terminator required";
```

Comment:
If we insist that this is an important check, it seems a bit jumpy to do the check at the beginning, after which we iterate over the elements and arrive exactly at that element - and just silently skip it.
Alternatively, we could check for `\0` when we get there, after we've processed the rest of the chars, i.e.
```C++
    assert(!hex_str[i]);
    return rv;
```

Reply:
I see what you mean and tried it out, but ended up preferring input validation at the top of the function.